[JIT] Make create autodiff subgraphs do in place updates to aliasDb

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42141 [JIT] Make create autodiff subgraphs do in place updates to aliasDb**
* #42140 refactor canonical ordering to also be able to do isAfter checks

Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

Commit:
Make create autodiff subgraphs do in place updates to aliasDb

[ghstack-poisoned]

Commit:
Update on "[JIT][WIP] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.



[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.



[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -12,6 +13,66 @@ namespace jit {
+    return a.user == b.user && a.offset == b.offset;
+  }
+
+  void copyAliasing(Node* merged_node, AliasDb& db) {
```
nitpick: should we just store `AliasDb` as a member of `ValueMapper` since we are already passing it in a c-tor which simplifies `copyAliasing` a little bit?

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -177,17 +264,18 @@ class SubgraphSlicer {
     if (shouldConsiderForMerge(consumer)) {
       if (consumer->kind() != prim::DifferentiableGraph) {
+        // ValueMapper preserves the aliasing information of the node's outputs
+        ValueMapper vm(consumer, aliasDb_, 0);
```
nitpick: we could probably just remember the number of the outputs of a producer (which we can get off the producer rather than pass both producer and the offset) since we will need to update aliasing for all of them and then in a consumer we could just update the aliasing information for the last  `consumer.outputs().size() - producer.outputs().size()` of the merged consumer?

eellison replied:
That's what I'm doing, see: https://github.com/pytorch/pytorch/pull/42141/files#diff-d35d4ac08adc5a54ab5b75564e9a1290R53

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -12,6 +13,66 @@ namespace jit {
+    auto g = n->owningGraph();
+    // temporary node to put the aliasing properties of the node before its
+    // merged and destroyed
+    placeholder_node_ = g->insertNode(g->create(prim::Uninitialized, 0));
```
nitpick: do we really need a node here? could we simply store all the values in a vector or will that confuse AA?

eellison replied:
Values are owned by the graph, if we store them in a vector but delete the node they will be dead references

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -12,6 +13,66 @@ namespace jit {
+    for (Value* v : new_outputs) {
+      auto maybe_last_use = firstOrLastUse(v, /*find_first*/ false);
+      // if it doesnt have a use it shouldnt have been added as output
+      TORCH_INTERNAL_ASSERT(maybe_last_use);
```
is this really true? could we have a node that produces two values but one of them isn't used?

eellison replied:
This is outputs of the autodiff subgraph, which are added individually. Even if there's a node with two outputs and one isn't used, the unused one will not be added an output to the autodiff subgraph

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -28,12 +89,54 @@ class SubgraphSlicer {
+        // Inlining nodes may cause some subexpression to come back in the
+        // subgraphs (for example, copying constants in repeatedly will generate
+        // redundant prim::Constants). Run CSE to clean them up.
+        EliminateCommonSubexpression(curNode->g(attr::Subgraph));
```
nitpick: should we move CSE into the next check? if we decide to inline it, the final CSE will clean all unmerged graphs up. In other words, we only need to clean a graph up if we aren't inlining it?

eellison replied:
This is existing logic in the file, I didn't change it. But yea, good point

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -200,37 +288,36 @@ class SubgraphSlicer {
 
+    ValueMapper vm(producer, aliasDb_, consumer->outputs().size());
     SubgraphUtils::mergeNodeIntoSubgraph(producer, consumer);
+    vm.copyAliasing(consumer, aliasDb_);
```
nitpick: I wonder if it makes sense to do `copyAliasing` when vm goes out of scope?

eellison replied:
What do you mean ?

Krovatkin commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -12,6 +13,66 @@ namespace jit {
+      TORCH_INTERNAL_ASSERT(i != last_uses_.size());
+      db.replaceWithNewValue(placeholder_node_->outputs().at(i), v);
+    }
+    placeholder_node_->destroy();
```
if `copyAliasing` is accidentally called twice `destroy` should throw an exception and not crash, right?

eellison replied:
If it's accidentally called twice we have other problems - this isn't an exposed API, this is only used within this file.

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]

Commit:
Update on "[JIT] Make create autodiff subgraphs do in place updates to aliasDb"


Update alias db in-place instead of having to construct alias db from scratch on each change, causing O(n^2) behavior. 

Description from https://github.com/pytorch/pytorch/pull/37106 holds pretty well: 
"""
Recomputing the aliasdb on every fusion iteration + in every subblock
is hugely expensive. Instead, update it in-place when doing fusion.

The graph fuser pass operates by pushing nodes into a fusion group. So
we start with

`x, y = f(a, b, c)`

and end with:
```
x_out, y_out = prim::fusionGroup(a, b, c)
   x_in, y_in = f(a_in, b_in, c_in)
   -> x_in, y_in
```

We destroy the x and y Value*s in the process. This operation is
easy to express as an update to the aliasDb--x_out just takes on all
the aliasing information x used to have. In particular, since we know
f and prim::fusionGroup are purely functional, we don't have to mess
with any write information.
"""

The one difficulty here is mapping x, y to x_out, y_out is not trivial in merging nodes into the autodiff subgraph node. 
There are a few options:
- attempt to make all subgraph utils & ir cloning logic update a map
- mirror the subgraph utils implementation in create_autodiff_subgraph
- uniquely map x, y and x_in, y_in so you can back out the correspondence.

I went with the third option. 

This shouldn't affect the results of the pass at all. LMK if you think there's anything else I should be doing to test, I was thinking about maybe exposing an option to run create autodiff subgraphs without the post processor and check that the alias db was correctly updated.

Differential Revision: [D22798377](https://our.internmc.facebook.com/intern/diff/D22798377)

[ghstack-poisoned]
